primitives/state-machine/src/trie_backend.rs

```diff
@@ -19,12 +19,104 @@
 
 use crate::{
 	trie_backend_essence::{TrieBackendEssence, TrieBackendStorage},
-	Backend, StorageKey, StorageValue,
+	Backend, ExecutionError, StorageKey, StorageValue,
 };
 use codec::Codec;
-use hash_db::Hasher;
+use hash_db::{HashDBRef, Hasher, EMPTY_PREFIX};
 use sp_core::storage::{ChildInfo, StateVersion};
 use sp_std::vec::Vec;
+#[cfg(feature = "std")]
+use sp_trie::{cache::LocalTrieCache, recorder::Recorder};
+use sp_trie::{MemoryDB, StorageProof};
+
+/// Builder for creating a [`TrieBackend`].
+pub struct TrieBackendBuilder<S: TrieBackendStorage<H>, H: Hasher> {
+	storage: S,
+	root: H::Out,
+	#[cfg(feature = "std")]
+	recorder: Option<Recorder<H>>,
+	#[cfg(feature = "std")]
+	cache: Option<LocalTrieCache<H>>,
+}
+
+impl<S, H> TrieBackendBuilder<S, H>
+where
+	S: TrieBackendStorage<H>,
+	H: Hasher,
+{
+	/// Create a new builder instance.
+	pub fn new(storage: S, root: H::Out) -> Self {
+		Self {
+			storage,
+			root,
+			#[cfg(feature = "std")]
+			recorder: None,
+			#[cfg(feature = "std")]
+			cache: None,
```

Comment:
could use cache in no_std (eg a local_value only cache, but may not be a good idea to use more wasm memory).

Reply:
I have some really good plans for the wasm side. However, this will come later. In the end the parachain validation is suffering as well from constant decoding for nodes.